# Выпускной проект

**SQL project**

## Сбор первичной информации о проекте

### Описание проекта

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

### Цель проекта

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

### Задачи проекта

1. Посчитайте, сколько книг вышло после 1 января 2000 года;
2. Для каждой книги посчитайте количество обзоров и среднюю оценку;
3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

### Описание данных

**Таблица `books`** - Содержит данные о книгах:
- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`** - Содержит данные об авторах:
- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`** - Содержит данные об издательствах:
- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`** - Содержит данные о пользовательских оценках книг:
- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`** - Содержит данные о пользовательских обзорах на книги:
- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

### Структура базы данных

![fig](https://i.ibb.co/d5vwhhb/Untitled-33.png)

## Основной проект

### Загрузка и обзор данных из базы данных:

Для доступа к базе данных была использована соответствующая [инструкция][1]

[1]:https://disk.yandex.ru/i/cwJdOjH0foH1Mw

In [1]:
#!/usr/bin/python
# -*- coding: utf-8 -*-

# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

# Задаём параметры подключения к БД,
db_config = {'user': 'praktikum_student'                        , # имя пользователя
             'pwd' : 'Sdf4$2;d-d30pp'                           , # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net', # host
             'port': 6432                                       , # порт подключения
             'db'  : 'data-analyst-final-project-db'            } # название базы данных

# Формируем строку соединения с БД.
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'] ,
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db']  )

# Подключаемся к БД.
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

Выполню проверку удалось ли мне подключится к базе данных и какие данные есть у нас в файлах.

In [2]:
# Формируем sql-запрос.
query = ''' SELECT *
            FROM books
        '''
# Выполняем запрос и сохраняем результат выполнения в DataFrame.
df_books = pd.io.sql.read_sql(query, con = engine)
       
# Формируем sql-запрос.
query = ''' SELECT *
            FROM authors
        '''
# Выполняем запрос и сохраняем результат выполнения в DataFrame.
df_authors = pd.io.sql.read_sql(query, con = engine)

# Формируем sql-запрос.
query = ''' SELECT *
            FROM publishers
        '''
# Выполняем запрос и сохраняем результат выполнения в DataFrame.
df_publishers = pd.io.sql.read_sql(query, con = engine)

# Формируем sql-запрос.
query = ''' SELECT *
            FROM ratings
        '''
# Выполняем запрос и сохраняем результат выполнения в DataFrame.
df_ratings = pd.io.sql.read_sql(query, con = engine)

# Формируем sql-запрос.
query = ''' SELECT *
            FROM reviews
        '''
# Выполняем запрос и сохраняем результат выполнения в DataFrame.
df_reviews = pd.io.sql.read_sql(query, con = engine)

In [3]:
# Get common information about datasets
def get_info(df, df_name):
    '''
    df : Dataframe,
        The dataset with data for the project
    df_name : Objects
        The name of the dataset for information
        
    Return:
        There are no return fields
    '''
    print(f'Common information about - {df_name}')
    display(df.head())
    df.info()
    print(df.columns, '\n')   
    print(f'Numbers of NaN values in the dataset - {df_name}', '\n')
    print(df.isnull().sum())
    print(f'Numbers of duplicates (explicit)in the dataset - {df_name}', '\n')
    print(df.duplicated().sum())

In [4]:
# Просмотр общей информации о файле
print(get_info(df_books, 'df_books')) 

Common information about - df_books


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB
Index(['book_id', 'author_id', 'title', 'num_pages', 'publication_date',
       'publisher_id'],
      dtype='object') 

Numbers of NaN values in the dataset - df_books 

book_id             0
author_id           0
title               0
num_pages           0
publication_date    0
publisher_id        0
dtype: int64
Numbers of duplicates (explicit)in the dataset - df_books 

0
None


**Вывод:** Таблица `books` успешно загружена, данные не содержат дубликатов и пропусков. При дальнейшей работе обратить внимание, что столбец *publication_date* имеет строковый тип. Возможно, что в запросе нужно будет изменить временной тип.

In [5]:
# Просмотр общей информации о файле
print(get_info(df_authors, 'df_authors'))

Common information about - df_authors


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB
Index(['author_id', 'author'], dtype='object') 

Numbers of NaN values in the dataset - df_authors 

author_id    0
author       0
dtype: int64
Numbers of duplicates (explicit)in the dataset - df_authors 

0
None


**Вывод:** Таблица `authors` успешно загружена, данные не содержат дубликатов и пропусков. Типы данных соответствуют заданным параметрам.

In [6]:
# Просмотр общей информации о файле
print(get_info(df_publishers, 'df_publishers'))

Common information about - df_publishers


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB
Index(['publisher_id', 'publisher'], dtype='object') 

Numbers of NaN values in the dataset - df_publishers 

publisher_id    0
publisher       0
dtype: int64
Numbers of duplicates (explicit)in the dataset - df_publishers 

0
None


**Вывод:** Таблица `publishers` успешно загружена, данные не содержат дубликатов и пропусков. Типы данных соответствуют заданным параметрам.

In [7]:
# Просмотр общей информации о файле
print(get_info(df_ratings, 'df_ratings'))

Common information about - df_ratings


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB
Index(['rating_id', 'book_id', 'username', 'rating'], dtype='object') 

Numbers of NaN values in the dataset - df_ratings 

rating_id    0
book_id      0
username     0
rating       0
dtype: int64
Numbers of duplicates (explicit)in the dataset - df_ratings 

0
None


**Вывод:** Таблица `ratings` успешно загружена, данные не содержат дубликатов и пропусков. Типы данных соответствуют заданным параметрам.

In [8]:
# Просмотр общей информации о файле
print(get_info(df_reviews, 'df_reviews'))

Common information about - df_reviews


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB
Index(['review_id', 'book_id', 'username', 'text'], dtype='object') 

Numbers of NaN values in the dataset - df_reviews 

review_id    0
book_id      0
username     0
text         0
dtype: int64
Numbers of duplicates (explicit)in the dataset - df_reviews 

0
None


**Вывод:** Таблица `reviews` успешно загружена, данные не содержат дубликатов и пропусков. Типы данных соответствуют заданным параметрам.

### Основная часть работы:

#### Запрос 1

***Посчитайте, сколько книг вышло после 1 января 2000 года.***

In [9]:
# Формируем sql-запрос.
query = ''' SELECT COUNT(title)
            FROM books
            WHERE CAST(publication_date AS date) >= '2000-01-02';
        '''
# Выполняем запрос и сохраняем результат выполнения в DataFrame.
report = pd.io.sql.read_sql(query, con = engine)
report

count
0    819

*Вывод:* С 1 января 2000 года было выпущено 819 книг

#### Запрос 2

***Для каждой книги посчитайте количество обзоров и среднюю оценку.***

In [10]:
query = '''SELECT b.title,
                  COUNT(DISTINCT rev.text), 
                  AVG(rat.rating)
           FROM books AS b
           LEFT JOIN reviews AS rev ON b.book_id = rev.book_id
           LEFT JOIN ratings AS rat ON b.book_id = rat.book_id
           GROUP BY b.title
           ORDER BY COUNT(DISTINCT rev.text) DESC;   
        '''
# Выполняем запрос и сохраняем результат выполнения в DataFrame.
report = pd.io.sql.read_sql(query, con = engine)
report.columns = ['book', 'reviews_numbers', 'mean_rating']

print('Top 15 books with maximum reviews')
display(report.head(15))

print('Top 5 books with minimum reviews')
display(report.tail(5))

Top 15 books with maximum reviews


book  reviews_numbers  \
0                                 Memoirs of a Geisha                8   
1                             Twilight (Twilight  #1)                7   
2   Harry Potter and the Chamber of Secrets (Harry...                6   
3                                    The Glass Castle                6   
4                                     Eat  Pray  Love                6   
5   The Curious Incident of the Dog in the Night-Time                6   
6                                 Water for Elephants                6   
7                           Outlander (Outlander  #1)                6   
8                                            The Road                6   
9                                       The Alchemist                6   
10             The Da Vinci Code (Robert Langdon  #2)                6   
11                             The Catcher in the Rye                6   
12                The Hobbit  or There and Back Again                6   
13  The Lightning Thief (Percy Jackson and the Oly...                6   
14  Harry Potter and the Prisoner of Azkaban (Harr...                6   

    mean_rating  
0      4.138462  
1      3.662500  
2      4.287500  
3      4.206897  
4      3.395833  
5      4.081081  
6      3.977273  
7      4.125000  
8      3.772727  
9      3.789474  
10     3.830508  
11     3.825581  
12     4.125000  
13     4.080645  
14     4.414634

Top 5 books with minimum reviews


book  reviews_numbers  \
994    Anne Rice's The Vampire Lestat: A Graphic Novel                0   
995                            The Natural Way to Draw                0   
996   The Cat in the Hat and Other Dr. Seuss Favorites                0   
997                          Essential Tales and Poems                0   
998  Disney's Beauty and the Beast (A Little Golden...                0   

     mean_rating  
994     3.666667  
995     3.000000  
996     5.000000  
997     4.000000  
998     4.000000

*Вывод:* Больше всего обзоров от пользователей получили такие книги, как *Memoirs of a Geisha*, *Twilight (Twilight #1)*, *Harry Potter*. Также в нашей базе данных встречаются книги, которые не обладают обзором от пользователей, что в принципе абсолютно логично, поскольку это могут быть книги более узкоспециализированные. 

#### Запрос 3

***Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры.***

In [11]:
query = '''SELECT publisher,
                  COUNT(b.title)         
           FROM publishers AS publ
           INNER JOIN books AS b ON publ.publisher_id = b.publisher_id
           WHERE b.num_pages > 50
           GROUP BY publisher
           ORDER BY COUNT(b.title) DESC;
        '''
# Выполняем запрос и сохраняем результат выполнения в DataFrame.
report = pd.io.sql.read_sql(query, con = engine)
report.columns = ['publisher_name', 'book_numbers']

print('Top 10 publishers:')
report.head(10)

Top 10 publishers:


publisher_name  book_numbers
0             Penguin Books            42
1                   Vintage            31
2  Grand Central Publishing            25
3          Penguin Classics            24
4          Ballantine Books            19
5                    Bantam            19
6                   Berkley            17
7        St. Martin's Press            14
8             Berkley Books            14
9                     Delta            13

*Вывод:* Больше всего книг толще 50 страниц было отпечатано издательством ***Penguin Books***.

#### Запрос 4

***Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками.***

In [12]:
query = '''SELECT top_score.author             AS author,
                  AVG(top_score.avg_rat)       AS avg_rating
                  
           FROM (SELECT a.author,  
                        b.book_id,
                        AVG(r.rating)          AS avg_rat,
                        COUNT(r.rating) 
                 FROM authors                  AS a   
                 INNER JOIN books              AS b ON b.author_id = a.author_id
                 INNER JOIN ratings            AS r ON r.book_id   = b.book_id
                 GROUP BY a.author, b.book_id
                 HAVING COUNT(r.rating) >= 50) AS top_score
           GROUP BY author
           ORDER BY avg_rating DESC
        '''

# Выполняем запрос и сохраняем результат выполнения в DataFrame.
report = pd.io.sql.read_sql(query, con = engine)
print('Top 1 author with the hightest score:')
report.head(1)

Top 1 author with the hightest score:


author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.283844

*Вывод:*  J.K. Rowling имеет наивысший рейтинг среди книг имеющих с 50 и более оценками.

#### Запрос 5

***Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.***

In [13]:
query = '''SELECT avg(count) 
           FROM (SELECT COUNT(text)
                 FROM reviews as rev
                 FULL OUTER JOIN (SELECT username,
                                  COUNT(rating_id) as urating
                                  FROM ratings
                                  GROUP BY username) AS user_rat ON user_rat.username = rev.username
                  WHERE urating > 50
                  GROUP BY rev.username) as count;                                  
        '''
# Выполняем запрос и сохраняем результат выполнения в DataFrame.
report = pd.io.sql.read_sql(query, con = engine)
report.columns = ['mean_reviews'] 
report  

mean_reviews
0     24.333333

*Вывод:* В среднем пользователи, которые оставили больше 50 отзывов пишут по 24 обзора. 

## Общий вывод:

Для запуска книжного магазина нам точно нужно, чтобы у нас были книги из вселенной Гарри Поттера (авт.: J.K. Rowling) и Властелина Колец (авт. J.R.R. Tolkien). 